Set up color maps for Decanter and refactor main nav colors to refer to color maps

yvonnetangsu opened the pull request:
# READY FOR REVIEW

# Summary
- Set up color maps for Decanter - a main one `$core-colors` and one for main-nav `$main-nav-colors`. There is already an existing one called `$alert-colors`.
- Add a function `color` to get colors from color maps.
- Refactor main-nav and a few other components to pull colors from the color map
- Also included issue 422 - changed all reference to $color-dark-red to $color-bright-red for accessibility

# Needed By (Date)
- Whenever

# Urgency
- Not urgent

# Steps to Test

1. Pull this branch
2. See that the main nav, buttons, brand bar, logo, global footer, alerts, links and all variants still have the correct colors.
3. Look at the new color maps in utilities/variables and the function `color`.
4. See that all places in main-nav with $color-dark-red have been changed to $color-bright-red.

# Affected Projects or Products
- Decanter

# Associated Issues and/or People
- https://github.com/SU-SWS/decanter/issues/386
- https://github.com/SU-SWS/decanter/issues/420
- https://github.com/SU-SWS/decanter/issues/422

sherakama commented on core/src/scss/components/alert/_alert--error.scss:
```diff
@@ -2,7 +2,7 @@
 // Error alert.
 .su-alert--error {
-  background-color: map-get($alert-colors, 'error');
+  background-color: color(error, $alert-colors);
```
Oooh, I like this. Super clean.

sherakama commented on core/src/scss/components/main-nav/_main-nav--dark.scss:
```diff
@@ -1,128 +1,137 @@
-          &::before {
-            background-color: $color-white;
-          }
+.su-main-nav--dark {
```
Yes! less nesting FTW.

sherakama commented on core/src/scss/components/main-nav/_main-nav--dark.scss:
```diff
@@ -1,128 +1,137 @@
+      &:hover,
+      &:focus {
+        &::before {
+          background-color: color(border--hover, $main-nav-colors, dark);
```
This is a little less clear than the two parameter function. Instead of having nested themes, why wouldn't you just pass in a different color map? eg:

`background-color: color(border--hover, $main-nav-colors-dark);`

sherakama commented on core/src/scss/utilities/functions/_color.scss:
```diff
@@ -0,0 +1,53 @@
+// Style guide: Functions.Color
+//
+
+@function color($color, $colormap: $core-colors, $theme: default) {
```
I think the third parameter (theme) could be removed in favor of simply passing in a different colormap.

sherakama replied:
Also, this is a great interface for color changes. YES THANK YOU!

yvonnetangsu commented:
> I like this, I like this a lot. My only suggestion is to simplify the color function by removing the third parameter but I am open for influence or feedback from others on this.

Awesome 😄 Actually you're right. I'm not sure there's any real reason to use the nested maps. @JBCSU , are you ok with @sherakama 's suggestion to not have nested maps?

sherakama commented:
We should only be rendering one 'theme' at a time. To do that, we will need some sort of toggle and that will likely be in some sort of variable changes. If I were to create themes today I would create a new partial file with all of the variable changes and include it in my rendering of css. We could package themes in this project by adding those partial files as new entry points in webpack but I think I would rather provide documentation and an examples file instead of having to support multiple themes per component. We could unofficially share those theme partials across our projects through gists or something else.

yvonnetangsu commented:
@sherakama @JBCSU I've removed the nesting in the color maps and changed the function @color to get rid of the $theme parameters. Also, I've changed all reference to $color-dark-red to $color-bright-red per @kerri-augenstein 's suggestion (https://github.com/SU-SWS/decanter/issues/422). Could you please give it another look? Thanks!

sherakama commented on core/src/scss/components/main-nav/_main-nav--dark.scss:
```diff
@@ -1,128 +1,139 @@
-            background-color: $color-white;
-          }
+.su-main-nav--dark {
+  $_color-map: $main-nav-dark-colors;
```
Just in case this variable name is used somewhere else, can you prefix it with the component name. 

eg: `$main-nav-color-map`

sherakama commented on core/src/scss/components/main-nav/_main-nav--light.scss:
```diff
@@ -1,189 +1,233 @@
-.su-main-nav {
-  &.su-main-nav--light {
+.su-main-nav--light {
+  $_color-map: $main-nav-light-colors;
```
Please namespace this variable.

yvonnetangsu replied:
Changed from $_color-map to $_su-main-nav-color-map (combining both @sherakama and @JBCSU 's suggestions)

sherakama commented on core/src/scss/utilities/functions/_color.scss:
```diff
@@ -0,0 +1,34 @@
+      $color: map-get($colormap, $color);
+    }
+    @else {
+      @warn "Check that `#{$color}` is in color map."
```
Why not error out? I think we would want to have this break if a color is not available.

sherakama commented on core/src/scss/utilities/functions/_color.scss:
```diff
@@ -0,0 +1,34 @@
+      $color: map-get($core-colors, $color);
+    }
+    @else {
+      @warn "Check that `#{$color}` is in color map $core-colors."
```
Why not error out? I think we would want to have this break if a color is not available.

yvonnetangsu replied:
@sherakama , I changed it to @error but when I tried putting in a wrong color name, it doesn't show my custom error - any idea why?

sherakama replied:
This may be something to do with the compiler and not the sass itself. The way we build in webpack may be hiding the error string.

sherakama commented on core/src/scss/utilities/variables/components/_main-nav.scss:
```diff
@@ -0,0 +1,151 @@
+//    link--mobile-expanded:                  $color-white,
+//    border--mobile-expanded:                transparent,
+//    border--mobile-active-expanded:         $color-white,
+//    light: (
```
These need to be pulled out and reflected with the new multiple variable option.

yvonnetangsu replied:
Ooops...meant to do that. I'll fix this and make the other changes. Thanks!

sherakama commented on core/src/scss/utilities/variables/components/_main-nav.scss:
```diff
@@ -0,0 +1,151 @@
+  border--mobile-active-expanded:         $color-white,
+) !default;
+
+$main-nav-light-colors: (
```
These should have their own KSS comment.

yvonnetangsu replied:
Added separate section with markup/comments for each variants

sherakama commented on core/src/scss/utilities/variables/core/_core-color-maps.scss:
```diff
@@ -0,0 +1,46 @@
+//
+// $core-colors
+//
+// SASS color map for the core Decanter theme. In the future, there might be more than one color theme.
```
I think you can remove the part about themes.

sherakama commented:
Not sure why CC is complaining about the line length. I'm pretty sure we have that set to 120.

yvonnetangsu commented:
> Not sure why CC is complaining about the line length. I'm pretty sure we have that set to 120.

Yeah, CC's been weird lately. It complains about the braces not on their own lines but they are :\
